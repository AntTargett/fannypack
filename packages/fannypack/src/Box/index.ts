export * from './Box';
export * from './styled';
