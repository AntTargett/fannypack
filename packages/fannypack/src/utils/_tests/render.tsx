import * as React from 'react';
import { render } from '@testing-library/react';
// @ts-ignore
import { Provider } from '../../Provider';

export default (Component: any, { theme = {} } = {}) => render(<Provider theme={theme}>{Component}</Provider>);
