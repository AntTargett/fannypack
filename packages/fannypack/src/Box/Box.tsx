import * as React from 'react';
import { Box as ReakitBox, BoxProps as ReakitBoxProps } from 'reakit';

import styled, { theme } from '../styled';
import { BoxThemeConfig, CSSProperties } from '../types';
import * as utils from '../utils';

export type LocalBoxProps = {
  use?: string | React.ComponentType<any>;
  children?: React.ReactNode | ((props: BoxProps) => React.ReactNode);
  /* Component-level theme overrides [Read more](TODO) */
  overrides?: BoxThemeConfig;
};
export type BoxProps = ReakitBoxProps & CSSProperties & LocalBoxProps;

export function useBoxProps(props: BoxProps | void) {
  const { use = undefined, ...restProps } = props || {};

  // Convert CSS props to a "style" prop & remove CSS props from DOM element.
  // Example input:
  // restProps = { color: 'red', backgroundColor: 'blue', isEnabled: true }
  //
  // Example output:
  // style = { color: 'red', backgroundColor: 'blue' }
  const style = utils.useStyle(restProps);

  // Since we now have a "style" prop, let's omit CSS props from the DOM element.
  const newProps = utils.omitCSSProps(restProps);

  return { use: use || StyledBox, style, ...utils.omitCSSProps(newProps) };
}

export function Box(props: BoxProps) {
  const { children, ...restProps } = props;
  const boxProps = useBoxProps(restProps);
  if (utils.isFunction(children)) {
    return <React.Fragment>{children(boxProps)}</React.Fragment>;
  }
  return (
    <StyledBox {...boxProps} as={boxProps.use !== StyledBox ? boxProps.use : undefined}>
      {children}
    </StyledBox>
  );
}

// TODO: Migrate to emotion, use shouldForwardProps
export const StyledBox = styled(ReakitBox)<BoxProps>`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme('Box.base')};
  }
`;
