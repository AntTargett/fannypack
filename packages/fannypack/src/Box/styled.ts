import { Box as ReakitBox } from 'reakit';
import styled, { theme } from '../styled';
import { BoxProps } from './Box';

export const StyledBox = styled(ReakitBox)`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme('Box.base')};
  }
`;
