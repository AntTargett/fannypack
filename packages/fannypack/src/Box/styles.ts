import _get from 'lodash/get';
import { css, cssClass } from '../styled';
import { theme } from '../utils';

export const style = styleProps => cssClass`
  && {
    ${styleProps.style};
  }
`;

export const Box = styleProps => cssClass`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  ${buildVisibleAttributes(styleProps)};

  & {
    ${theme('Box.base')(styleProps)};
  }
`;

function buildVisibleAttributes(props: { hiddenBreakpoint?: string; showBreakpoint?: string }) {
  let breakpoint = props.hiddenBreakpoint || props.showBreakpoint;
  if (!breakpoint) return;

  let key: string | undefined;
  if (props.hiddenBreakpoint && breakpoint.includes('max')) {
    key = 'max-width';
  } else if (props.hiddenBreakpoint && breakpoint.includes('min')) {
    key = 'min-width';
  } else if (props.showBreakpoint && breakpoint.includes('max')) {
    key = 'min-width';
  } else if (props.showBreakpoint && breakpoint.includes('min')) {
    key = 'max-width';
  }

  let strippedBreakpoint = breakpoint;
  strippedBreakpoint = strippedBreakpoint.replace('max-', '');
  strippedBreakpoint = strippedBreakpoint.replace('min-', '');

  const minBreakpointValues: { [key: string]: number } = {
    mobile: 0,
    tablet: _get(props, 'theme.layout.mobileBreakpoint'),
    desktop: _get(props, 'theme.layout.tabletBreakpoint'),
    widescreen: _get(props, 'theme.layout.desktopBreakpoint'),
    fullHD: _get(props, 'theme.layout.widescreenBreakpoint')
  };
  let breakpointValue = _get(props, `theme.layout['${strippedBreakpoint}Breakpoint']`);
  if (props.hiddenBreakpoint && breakpoint.includes('min')) {
    breakpointValue = minBreakpointValues[strippedBreakpoint] + 1;
  }
  if (props.showBreakpoint && breakpoint.includes('min')) {
    breakpointValue = minBreakpointValues[strippedBreakpoint];
  }
  if (props.showBreakpoint && breakpoint.includes('max')) {
    breakpointValue = breakpointValue + 1;
  }

  if (!breakpoint.includes('min-') && !breakpoint.includes('max-')) {
    if (props.showBreakpoint) {
      return css`
        @media screen and (max-width: ${minBreakpointValues[breakpoint]}px) {
          display: none;
        }

        @media screen and (min-width: ${breakpointValue + 1}px) {
          display: none;
        }
      `;
    }
    return css`
      @media screen and (min-width: ${minBreakpointValues[breakpoint] + 1}px) and (max-width: ${breakpointValue}px) {
        display: none;
      }
    `;
  }
  return css`
    @media screen and (${key}: ${breakpointValue}px) {
      display: none;
    }
  `;
}
