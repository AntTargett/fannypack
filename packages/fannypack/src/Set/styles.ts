import { css, cssClass } from '../styled';
import { space, theme } from '../utils';

export const Set = styleProps => cssClass`
  ${
    styleProps.isVertical
      ? css`
          flex-direction: column;

          ${!styleProps.isFilled &&
            css`
              align-items: flex-start;
            `};
        `
      : css`
          flex-wrap: wrap;
          align-items: center;
          justify-content: flex-start;
          margin-left: -${space(styleProps.spacing)(styleProps)}rem;
          margin-top: -${space(styleProps.spacing)(styleProps)}rem;
        `
  };

  & > * {
    ${
      styleProps.isVertical
        ? css`
            &:not(:last-child) {
              margin-bottom: ${space(styleProps.spacing)(styleProps)}rem;
            }
          `
        : css`
            margin-left: ${space(styleProps.spacing)(styleProps)}rem;
            margin-top: ${space(styleProps.spacing)(styleProps)}rem;
          `
    };

    ${theme('Set.child')(styleProps)};
  }

  & {
    ${theme('Set.base')(styleProps)};
  }
`;
