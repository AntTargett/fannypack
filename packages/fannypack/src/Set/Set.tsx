import { Box as ReakitBox } from 'reakit';

import { useClassName, createComponent, createElement } from '../utils';
import { Box, BoxProps } from '../Box';

import * as styles from './styles';

export type LocalSetProps = {
  isFilled?: boolean;
  isVertical?: boolean;
  spacing?: string;
};
export type SetProps = BoxProps & LocalSetProps;

function useProps(props: Partial<SetProps> = {}) {
  const boxProps = Box.useProps(props);

  const className = useClassName({
    style: styles.Set,
    styleProps: props,
    prevClassName: boxProps.className
  });

  return { ...boxProps, className };
}

export const Set = createComponent<SetProps>(
  props => {
    const setProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: setProps });
  },
  {
    attach: {
      defaultProps: {
        isFilled: false,
        isVertical: false,
        spacing: 'minor-2'
      },
      useProps
    },
    themeKey: 'Set'
  }
);
