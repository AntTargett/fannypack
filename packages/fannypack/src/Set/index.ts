import Set from './Set';

export default Set;

export * from './Set';
export { default as Set } from './Set';
