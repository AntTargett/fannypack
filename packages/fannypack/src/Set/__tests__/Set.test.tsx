import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Box } from '../../Box';
import { Set } from '../Set';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Set>
        <Box>hello</Box>
        <Box>world</Box>
      </Set>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with CSS props', () => {
    const { container } = render(
      <Set color="primary">
        <Box>hello</Box>
        <Box>world</Box>
      </Set>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      const { container } = render(
        <Set use="div">
          <Box>hello</Box>
          <Box>world</Box>
        </Set>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Set props', () => {
      const { result } = renderHook(() => Set.useProps());
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      const { container } = render(
        <Set>
          {SetProps => (
            <div {...SetProps}>
              <Box>hello</Box>
              <Box>world</Box>
            </div>
          )}
        </Set>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('overrides', () => {
  it('Set.base should render correctly', () => {
    const { container } = render(
      <Set overrides={{ Set: { base: { backgroundColor: 'red' } } }}>
        <Box>hello</Box>
        <Box>world</Box>
      </Set>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('Set.base should render correctly', () => {
    const { container } = render(
      <Set>
        <Box>hello</Box>
        <Box>world</Box>
      </Set>,
      {
        theme: { Set: { base: { backgroundColor: 'red' } } }
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Set.child should render correctly', () => {
    const { container } = render(
      <Set>
        <Box>hello</Box>
        <Box>world</Box>
      </Set>,
      {
        theme: { Set: { child: { backgroundColor: 'red' } } }
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    const { container } = render(
      <Set>
        <Box>hello</Box>
        <Box>world</Box>
      </Set>,
      {
        theme: { Set: { defaultProps: { className: 'test' } } }
      }
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
