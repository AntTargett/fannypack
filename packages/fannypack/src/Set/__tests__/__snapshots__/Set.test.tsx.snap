// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

<div
  class="emotion-4 emotion-0 emotion-1"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`composition hook should return with Set props 1`] = `
Object {
  "className": "css-49sw4z css-1pwrqjj css-1ccgglo",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

<div
  class="emotion-4 emotion-0 emotion-1"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

<div
  class="emotion-4 emotion-0 emotion-1 test"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`overrides Set.base should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-4 {
  background-color: red;
}

<div
  class="emotion-4 emotion-0 emotion-1"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

<div
  class="emotion-4 emotion-0 emotion-1"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`props should render correctly with CSS props 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-6.emotion-6.emotion-6 {
  color: #3722ed;
}

<div
  class="emotion-4 emotion-0 emotion-6"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`theming Set.base should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
}

.emotion-4 {
  background-color: red;
}

<div
  class="emotion-4 emotion-0 emotion-1"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;

exports[`theming Set.child should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

.emotion-4 {
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  margin-left: -0.5rem;
  margin-top: -0.5rem;
}

.emotion-4 > * {
  margin-left: 0.5rem;
  margin-top: 0.5rem;
  background-color: red;
}

<div
  class="emotion-4 emotion-0 emotion-1"
>
  <div
    class="emotion-0 emotion-1"
  >
    hello
  </div>
  <div
    class="emotion-0 emotion-1"
  >
    world
  </div>
</div>
`;
